internal/order/order.go

```diff
@@ -0,0 +1,88 @@
+package order
+
+import "math/rand"
+
+// An Order holds values to return in a randomized round-based sequence such
+// that each value is returned once per round. The sequence is shuffled after
+// each round. The zero value of type Order is an empty Order ready for use.
+type Order[T comparable] struct {
+	a    []T
+	next int
+}
+
+// Next returns the next value in the Order, shuffling first if necessary. It
+// returns the zero value of type T if the Order is empty.
+func (o *Order[T]) Next() T {
+	var t T
+	if len(o.a) == 0 {
+		return t
+	}
+	if o.next == len(o.a) {
+		o.next = 0
+		rand.Shuffle(len(o.a), o.swap)
+	}
+	t = o.a[o.next]
+	o.next++
+	return t
+}
+
+// Add inserts t into a random position in the Order. Depending on where it is
+// inserted, t may or may not be returned in the current round.
+func (o *Order[T]) Add(t T) {
+	o.addAt(t, rand.Intn(len(o.a)+1))
+}
+
+// addAt inserts t at index k, which must be in the range [0, len(o.a)].
+func (o *Order[T]) addAt(t T, k int) {
```

Comment:
This method is only called from one place. Maybe it should be inlined?

[John Carmack had some thoughts on inlined code](http://number-none.com/blow/john_carmack_on_inlined_code.html) that I find fairly persuasive.

Reply:
I agree in principle, but I wanted to be able to test it independently of the randomization.

Reply:
Right on.